### ISSUE TITLE
Could anybody help check this problem? I've been stuck in two days and didn't find the solution.

### ISSUE THREAD

**Issue by glhfgg1024** [tensorflow/tensorflow#19002]
I'm using the tf.Dataset API as follows: Suppose I have 10 `*.tfrecords` files (named `trn-001.tfrecords`,`trn-002.tfrecords`,etc.). There are 128 samples in each of these tfrecords. Suppose the samples in the first 5 tfrecords belong to class 0 and the remaining 5 tfrecords belong to class 1. In my case, the data provider should first random select 2 tfrecords for each class from those 5 tfrecords, then select fixed number of samples from each of the selected tfrecords and combine them to generate one balanced batch data to feed into the network for training. For example, for class 0, select `trn-000.tfrecords` and `trn-001.tfrecords`; for class 1, select `trn-006.tfrecords` and `trn-009.tfrecords`; then for each of the four selected tfrecords, randomly select 1 sample from each of them and combine the `1+1+1+1` samples to generate one batch data. My script to initialize the dataset operations is like this:

    trn0_filenames = ['trn-000.tfrecords','trn-001.tfrecords','trn-002.tfrecords','trn-003.tfrecords','trn-004.tfrecords']
    trn1_filenames = ['trn-005.tfrecords','trn-006.tfrecords','trn-007.tfrecords','trn-008.tfrecords','trn-009.tfrecords']

    half_batch_size = int(BATCH_SIZE / 2)
    trn0_count = len(trn0_filenames)
    trn1_count = len(trn1_filenames)
    features0_ops = []
    labels0_ops = []
    for trn0_filename in trn0_filenames:
        feature_op, label_op, = dataset_input_from_tfrecords([trn0_filename], batch_size=1, num_epochs=50000, shuffle=True, image_mean=None)
        features0_ops.append(feature_op)
        labels0_ops.append(label_op)
    
    features1_ops = []
    labels1_ops = []
    for trn1_filename in trn1_filenames:
        feature_op, label_op = dataset_input_from_tfrecords([trn1_filename], batch_size=1, num_epochs=50000, shuffle=True, image_mean=None)
        features1_ops.append(feature_op)
        label_ops.append(label_op)

Then during training, the script to generate the batch data is like this:

    features = np.zeros((BATCH_SIZE, HEIGHT, WIDTH, 3), dtype=np.float32)
    labels = np.zeros((BATCH_SIZE,), dtype=np.int64)
    
    while not coord.should_stop():
        ind0 = np.random.choice(trn0_count, half_batch_size)
        ind1 = np.random.choice(trn1_count, half_batch_size)
        batch_idx = 0
        for idx in ind0:
            temp1,temp2 = sess.run([features0_ops[idx],
                                                labels0_ops[idx]])
            features[batch_idx] = temp1[0]
            labels[batch_idx] = temp2[0]
            batch_idx+=1
        for idx in ind1:
            temp1,temp2 = sess.run([features1_ops[idx],
                                    labels1_ops[idx]])
            features[batch_idx] = temp1[0]
            labels[batch_idx] = temp2[0]
            batch_idx+=1

and then feed the batch data into network for training:

    summary,_,loss_total,loss,loss_reg = sess.run([summary_op, train_op, 
                                   loss_total_op,loss_op,loss_reg_op], 
                                  feed_dict={x_tensor: features, y_tensor: labels})

The problem is: during training, I found that the memory usage of the process is consistently increasing. Please check the following screenshot. The process 3208 is the running process. At first, the memory usage is about `10%`, but at the moment when I taken this picture, the memory usage has increased to `41.6%`. So my question is: is this a memory problem? What's wrong? I cannot find the problem. If the memory usage is increasing to some certain amount, the computer will be halt down then I need to restart.

[![memory usage][1]][1]

  [1]: https://i.stack.imgur.com/WaPkI.png

**Comment by mrry** [comment]
Can you please share the code for the `dataset_input_from_tfrecords()` function, as well as what exact version of TensorFlow you're using?

**Comment by glhfgg1024** [comment]
Hi @mrry , if I use the ordinary version without randomly use the python `list` to save the ops (feature_op, label_op), and use the following:

    trn_filenames = trn0_filenames +trn1_filenames
    features_op, labels_op = dataset_input_from_tfrecords(filenames, batch_size=4, num_epochs=50000, shuffle=True, image_mean=None)

Then everything works fine, which means the `dataset_input_from_tfrecords` is fine. But in my case I want to make the class distribution balanced and just one sample from one tfrecords file. I'm not sure if the `Dataset` API has this functionality, thus I generated many Datasets operations and did the work like that described above.

**Comment by glhfgg1024** [comment]
@mrry Sorry, the problem may not lie in the Dataset API. Maybe lie in the following:  After one epoch, I conduct the prediction on validation set like this:

    if step%NUM_TRAIN_STEPS_ONE_EPOCH==0:
        val_features_op, val_labels_op = dataset_input_from_tfrecords(val_filenames, batch_size=1, num_epochs=50000, shuffle=True, image_mean=None)
        ....
        val_features, val_labels = sess.run([val_features_op, val_labels_op])
        ....
    
So the `val_features_op, val_labels_op` will add to the whole graph for many times during the loop, right? Maybe this is the problem.

**Comment by glhfgg1024** [comment]
@mrry , that's the problem. After I commented the validation code, the memory usage is reasonable during training. Sorry for disturbing you!

**Comment by mrry** [comment]
Thanks for confirming! Yes, that call to `dataset_input_from_tfrecords()` inside the training loop would cause the graph to grow.

**Comment by glhfgg1024** [comment]
@mrry , thanks a lot for your confirmation!
